File: backend/app/schemas/inventory.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class CurrentStockItem(BaseModel):
    inv_current_id: int
    item_id: int
    item_nm: str
    itemtyp_nm: str
    lot: str
    qty: Decimal
    updated_at: datetime


class GrgiHistoryItem(BaseModel):
    inv_grgi_id: int
    item_id: int
    item_nm: str
    lot: str
    move_qty: Decimal
    qty: Decimal
    movetyps_nm: str
    actual_at: datetime
    created_at: datetime | None = None


class GrgiCreate(BaseModel):
    item_id: int = Field(gt=0)
    lot: str = Field(min_length=1, max_length=50)
    move_qty: Decimal = Field(gt=0)
    movetyps_id: int = Field(gt=0)
    actual_at: datetime


class MoveTypOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    movetyps_id: int
    movetyps_nm: str


class LotTraceCurrent(BaseModel):
    item_id: int
    item_nm: str
    itemtyp_nm: str
    lot: str
    qty: Decimal
    updated_at: datetime


class LotTraceHistory(BaseModel):
    inv_grgi_id: int
    item_nm: str
    movetyps_nm: str
    move_qty: Decimal
    qty: Decimal
    actual_at: datetime
    created_at: datetime | None


class LotTraceBalance(BaseModel):
    period_year_month: str
    item_nm: str
    lot: str
    beg_at: datetime
    beg_qty: Decimal
    qty: Decimal


class LotTraceResult(BaseModel):
    lot: str
    current: list[LotTraceCurrent]
    history: list[LotTraceHistory]
    balances: list[LotTraceBalance]


class BalanceItem(BaseModel):
    inv_balance_id: int
    period_year_month: str
    item_id: int
    item_nm: str
    lot: str
    beg_at: datetime
    beg_qty: Decimal
    qty: Decimal


class BalanceCreateResult(BaseModel):
    period_year_month: str
    rows_saved: int

File: frontend/src/types/inventory.ts
export interface CurrentStock {
  inv_current_id: number
  item_id: number
  location_id: number
  location_cd: string
  location_nm: string
  item_cd: string
  item_nm: string
  itemtyp_id: number
  itemtyp_nm: string
  lot: string
  gr_date: string | null
  qty: string | number
  updated_at: string
}

export interface GrgiHistory {
  inv_grgi_id: number
  item_id: number
  location_id: number
  location_cd: string
  location_nm: string
  item_nm: string
  lot: string
  move_qty: string | number
  qty: string | number
  movetyps_cd: string
  movetyps_nm?: string | null
  actual_at: string
  created_at?: string | null
}

export interface MoveTyp {
  movetyps_id: number
  movetyps_cd: string
  movetyps_nm?: string | null
}

export interface GrgiCreatePayload {
  item_id: number
  location_id: number
  lot: string
  move_qty: number
  movetyps_id: number
  actual_at: string
}

export interface LotTraceResult {
  lot: string
  current: {
    item_id: number
    location_id: number
    location_cd: string
    location_nm: string
    item_nm: string
    itemtyp_id: number
    itemtyp_nm: string
    lot: string
    qty: string | number
    updated_at: string
  }[]
  history: {
    inv_grgi_id: number
    item_id: number
    location_id: number
    location_cd: string
    location_nm: string
    item_nm: string
    movetyps_cd: string
    movetyps_nm?: string | null
    move_qty: string | number
    qty: string | number
    actual_at: string
    created_at: string | null
  }[]
  balances: {
    period_year_month: string
    item_id: number
    location_id: number
    location_cd: string
    location_nm: string
    item_nm: string
    lot: string
    beg_at: string
    beg_qty: string | number
    qty: string | number
  }[]
}

export interface BalanceItem {
  inv_balance_id: number
  period_year_month: string
  item_id: number
  location_id: number
  location_cd: string
  location_nm: string
  item_nm: string
  lot: string
  beg_at: string
  beg_qty: string | number
  qty: string | number
}

export interface LocationMovePayload {
  item_id: number
  from_location_id: number
  to_location_id: number
  lot: string
  qty: number
  actual_at: string
}
